### ISSUE TITLE
[Feature Request] Example notebook explaining how to use REMBO model

### ISSUE THREAD

**Issue by roshni-kamath** [facebook/Ax#264]
I am not sure how REMBO or REMBOStrategy is used with any acquisition function. Can an example be given so that its understood ?

**Comment by Balandat** [comment]
In a nutshell, since REMBO calls the `gen` function of its `BotorchModel` superclass [here](https://github.com/facebook/Ax/blob/master/ax/models/torch/rembo.py#L192), it uses whatever acquisition function is specified by the standard `BotorchModel`. This can be changed by passing in a custom callable via the `acqf_constructor` argument in the constructor (the default is `getNEI` from [here](https://github.com/facebook/Ax/blob/master/ax/models/torch/botorch_defaults.py#L149)). 

This BoTorch [tutorial](https://botorch.org/v/latest/tutorials/custom_acquisition) gives an example for how to specify such an alternate constructor.

You can then use the [`get_REMBO`](https://github.com/facebook/Ax/blob/2e5ac27846aca4f2fe292a563d49529513bccf8a/ax/modelbridge/strategies/rembo.py#L52-L76) convenience function, you just have to call `get_REMBO` with a `acqf_constructor=your_custom_constructor` kwarg, where `your_custom_constructor` is your callable with the same signature as that of `getNEI`.

I hope this helps.

**Comment by roshni-kamath** [comment]
Hi,
I am still not sure how to get that working. How do use the optimize loop with REMBO model and use either ExpectedImprovement/q-EI as acquisition function ?

**Comment by bletham** [comment]
@roshni-kamath To use REMBO you will need to use the `REMBOStrategy` generation strategy that is defined here:
https://github.com/facebook/Ax/blob/cbc5f6a0385a5232f0e79b8eb8f9cbb207292bd2/ax/modelbridge/strategies/rembo.py#L79

For instance, if you function had dimensionality `D=100` and you wanted to use REMBO with a 6-d embedding, you would do

```
from ax.modelbridge.strategies.rembo import REMBOStrategy
from ax.service.managed_loop import optimize

rembo_strategy = REMBOStrategy(D=100, d=6, init_per_proj=3)

# Optimize as usual with the managed loop API
best_parameters, values, experiment, model = optimize(
    parameters=parameters,
    experiment_name="name",
    objective_name="objective",
    evaluation_function=evaluation_function,  # The function to optimize
    minimize=True,
    total_trials=100,
    generation_strategy=rembo_strategy,
)
```
This will use REMBO with EI.

If you are interested in high-dimensional problems, I'd highly recommend checking out the ALEBO method which is also implemented in Ax. It is used in exactly the same way, and there is an example notebook showing how to use it here: https://github.com/facebookresearch/alebo/blob/master/quickstart.ipynb

It is an improvement on REMBO that we have found typically works much better for problems with low-dimensional linear structure. There is a paper describing the method here: https://arxiv.org/abs/2001.11659

**Comment by roshni-kamath** [comment]
Hi,
Thank you for the explanation. May I know the difference between REMBO model and REMBOStrategy?

**Comment by bletham** [comment]
Basically,
* A Model is something that takes in an experiment and the current set of data, and is able to generate new points and make predictions.
* A GenerationStrategy specifies which sequence of models to use when doing an optimization. So, a GenerationStrategy would specify something like "Do 5 iterations of random search, and then switch to a GP with EI".

In the case of REMBO, the REMBOStrategy specifies that we should first do `init_per_proj` iterations of random search within each linear subspace, and then switch to fitting a GP in the subspace and using EI.

If you just want to pass in training data and get a fitted model you could use the `get_REMBO` function that @Balandat linked to above, but I think you will find this a bit tricky with REMBO. The REMBO model can only be given data that were generated from within the linear embedding, so you can't create training data at random points like you normally would and pass them to the model. If you the REMBOStrategy, it will appropriately sample random points from within the linear subspace for the initialization for you. Also, REMBO as described in the paper uses multiple projections in parallel (4); this is another detail that REMBOStrategy is handling for you.

Basically, rather than trying to give a model training data, you should just give the REMBOStrategy the evaluation function and then it will be able to appropriately create the initial training data for you and run the whole optimization. https://github.com/facebookresearch/alebo/blob/master/quickstart.ipynb is an example of how to use a GenerationStrategy to optimize an arbitrary callable.

**Comment by roshni-kamath** [comment]
Thank you very much for your explanation. By model I meant the SingleTaskGP model, where I can pass train_X and train_Y. 

I started with BoTorch to perform Bayesiamn Optimization using GP but since I am dealing with 20D I am  planning to use REMBO. 
While using BoTorch, I followed the example, given as: 
https://github.com/pytorch/botorch/blob/master/tutorials/closed_loop_botorch_only.ipynb

I am trying to adapt the existing code for the REMBOStrategy. What I didn't understand is:  
- Like in BoTorch it is possible to train the GP model with initial training data, how do I achieve that in Ax/REMBO?
- How can I update the model with optimized and get new observation?
- After update training points, how can I reinitialize the models so they are ready for fitting on next iteration?

I apologize for theses dense questions, but I couldn't understand much from the documentation. 
Thank you in advance for any kind of help.

**Comment by bletham** [comment]
The main difference between something like SingleTaskGP and REMBO and that REMBO is not just a model; it's a procedure for the entire Bayes opt loop. It sounds like you want to take your Bayes opt loop and replace SingleTaskGP with REMBO. That won't be possible, because REMBO is a closed-loop method, not a model. In fact, the model in REMBO is actually SingleTaskGP; everything that makes it REMBO happens outside the model (projecting points up to the high-dimensional space; setting bounds in the low-dimensional space). 

To answer your specific questions,
* The GP used in REMBO is SingleTaskGP, you just need to pass in data in the low-dimensional space. Your initialization data need to all be generated in the low-dimensional space.
* That happens the same as in usual BO, and then after optimization you project points up to the high-dimensional space.
* Re-fit with the updated data in the low-dimensional space.

And recall that for this to be REMBO as described in the REMBO paper, you need to do this with 4 models in parallel (each with a separate projection and dataset).

BoTorch provides models and acquisition functions for you to construct a Bayes opt loop yourself. So you could use the SingleTaskGP and EI acquisition functions to construct a REMBO loop by following the paper.

Ax constructs the Bayes opt loop for you, and can do all of the projecting to the high-dimensional space and managing of multiple models for you. If you want to use REMBO without having to code up the REMBO loop yourself, you'll need to use Ax to do that. This tutorial should help you understand how to use Ax, at which point if you wanted to switch to using REMBO (or another high-dimensional BO method like ALEBO) it would be a 2-line change as I describe above: https://ax.dev/tutorials/gpei_hartmann_loop.html

**Comment by lena-kashtelyan** [comment]
Closing this issue, since it's covered by @bletham's response above and has been inactive for a while.
